frontend/apps/app/hooks/useNavigationGuard.ts

```diff
@@ -0,0 +1,26 @@
+import { useEffect } from 'react'
+
+export const useNavigationGuard = (beforeNavigate: () => void) => {
+  useEffect(() => {
+    const handleClick = (event: MouseEvent) => {
+      if (
+        event.target instanceof Element &&
+        event.target.closest('a:not([target="_blank"])')
+      ) {
+        beforeNavigate()
+      }
+    }
+
+    const handleBeforeUnload = () => {
+      beforeNavigate()
+    }
+
+    window.addEventListener('beforeunload', handleBeforeUnload)
+    window.addEventListener('click', handleClick, true)
+
+    return () => {
+      window.removeEventListener('beforeunload', handleBeforeUnload)
+      window.removeEventListener('click', handleClick, true)
+    }
+  }, [beforeNavigate])
+}
```

Comment:
This implementation does not support Next.js's `router.push()`. However, I determined it could be postponed for this specific use case. Currently, the following are supported:

- Browser back/forward navigation
- Reload
- Link transitions

To support `router.push()`, a library like https://github.com/LayerXcom/next-navigation-guard would be required.